thewalrus/_hermite_multidimensional.py

```diff
@@ -143,4 +148,160 @@ def hafnian_batched(A, cutoff, mu=None, rtol=1e-05, atol=1e-08, renorm=False, ma
     return hermite_multidimensional(
         -A, cutoff, y=mu, renorm=renorm, make_tensor=make_tensor, modified=True
     )
+
+
 # Note the minus signs in the arguments. Those are intentional and are due to the fact that Dodonov et al. in PRA 50, 813 (1994) use (p,q) ordering instead of (q,p) ordering
+
+
+@jit(nopython=True)
+def dec(tup: Tuple[int], i: int) -> Tuple[int, ...]:  # pragma: no cover
+    r"""returns a copy of the given tuple of integers where the ith element has been decreased by 1
+
+    Args:
+        tup (Tuple[int]): the given tuple
+        i (int): the position of the element to be decreased
+
+    Returns:
+        Tuple[int,...]: the new tuple with the decrease on i-th element by 1
+    """
+    copy = tup[:]
+    return tuple_setitem(copy, i, tup[i] - 1)
+
+
+@jit(nopython=True)
+def remove(
+    pattern: Tuple[int, ...]
+) -> Generator[Tuple[int, Tuple[int, ...]], None, None]:  # pragma: no cover
+    r"""returns a generator for all the possible ways to decrease elements of the given tuple by 1
+
+    Args:
+        pattern (Tuple[int, ...]): the pattern given to be decreased
+
+    Returns:
+        Generator[Tuple[int, Tuple[int, ...]], None, None]: the generator
+    """
+    for p, n in enumerate(pattern):
+        if n > 0:
+            yield p, dec(pattern, p)
+
+
+SQRT = np.sqrt(np.arange(1000))  # saving the time to recompute square roots
+
+
+def hermite_multidimensional_numba(R, cutoff, y, C = 1, dtype=np.complex128):
+    # pylint: disable=too-many-arguments
+    r"""Returns the multidimensional Hermite polynomials :math:`C*H_k^{(R)}(y)`.
+
+    Here :math:`R` is an :math:`n \times n` square matrix, and
+    :math:`y` is an :math:`n` dimensional vector. The polynomials are
+    parametrized by the multi-index :math:`k=(k_0,k_1,\ldots,k_{n-1})`,
+    and are calculated for all values :math:`0 \leq k_j < \text{cutoff}`,
+
+    Args:
+        R (array[complex]): square matrix parametrizing the Hermite polynomial
+        cutoff (int): maximum size of the subindices in the Hermite polynomial
+        y (vector[complex]): vector argument of the Hermite polynomial
+        C (complex): first value of the Hermite polynomials, the default value is 1
+        dtype (data type): Specifies the data type used for the calculation
+
+    Returns:
+        array[data type]: the multidimensional Hermite polynomials
+    """
+    n, _ = R.shape
+    if y.shape[0] != n:
+        raise ValueError("The matrix R and vector y have incompatible dimensions")
+    num_modes = len(y) // 2
+    array = np.zeros(((cutoff,) * (2 * num_modes)), dtype=dtype)
+    array[(0,) * (2 * num_modes)] = C
+    all_idx = product(range(cutoff), repeat=2 * num_modes)
+    next(all_idx)
+    for idx in all_idx:
```

Comment:
For now this works, but let's note somewhere that this is probably where we can improve the speed: this is a pure python loop over all the indices of the tensor.